async_simple/coro/Collect.h

```diff
@@ -123,25 +127,120 @@ struct CollectAnyAwaiter {
         }  // end for
     }
     auto await_resume() {
-        assert(_result != nullptr);
-        return std::move(*_result);
+        if constexpr (sizeof...(Function) == 0) {
+            assert(_result != nullptr);
+            return std::move(*_result);
+        }
     }
 
     std::vector<LazyType, InAlloc> _input;
     std::shared_ptr<ResultType> _result;
+    std::tuple<Function...> _callback;
+};
+
+template <typename... Ts>
+struct CollectAnyVariadicPairAwaiter {
+    using InputType = std::tuple<Ts...>;
+
+    CollectAnyVariadicPairAwaiter(Ts&&... inputs)
+        : _input(std::move(inputs)...), _result(nullptr) {}
+
+    CollectAnyVariadicPairAwaiter(InputType&& inputs)
+        : _input(std::move(inputs)), _result(nullptr) {}
+
+    CollectAnyVariadicPairAwaiter(const CollectAnyVariadicPairAwaiter&) =
+        delete;
+    CollectAnyVariadicPairAwaiter& operator=(
+        const CollectAnyVariadicPairAwaiter&) = delete;
+    CollectAnyVariadicPairAwaiter(CollectAnyVariadicPairAwaiter&& other)
+        : _input(std::move(other._input)), _result(std::move(other._result)) {}
+
+    bool await_ready() const noexcept {
+        return _result && _result->has_value();
+    }
+
+    void await_suspend(std::coroutine_handle<> continuation) {
+        auto promise_type =
+            std::coroutine_handle<LazyPromiseBase>::from_address(
+                continuation.address())
+                .promise();
+        auto executor = promise_type._executor;
+        auto event =
+            std::make_shared<detail::CountEvent>(std::tuple_size<InputType>());
+        auto result = std::make_shared<std::optional<size_t>>();
+        _result = result;
+
+        auto input = std::move(_input);
+
+        [&]<size_t... I>(std::index_sequence<I...>) {
+            (
+                [&]() {
+                    if (result->has_value()) {
+                        return;
+                    }
+
+                    auto& first = std::get<0>(std::get<I>(input));
+
+                    if (!first._coro.promise()._executor) {
+                        first._coro.promise()._executor = executor;
+                    }
+
+                    auto callback = std::move(std::get<1>(std::get<I>(_input)));
+                    first.start(
+                        [result, event, continuation,
+                         callback = std::move(callback)](auto&& res) mutable {
+                            auto count = event->downCount();
+                            if (count == std::tuple_size<InputType>() + 1) {
+                                callback(std::move(res));
+                                *result = I;
+                                continuation.resume();
+                            }
+                        });
+                }(),
+                ...);
+        }
+        (std::make_index_sequence<sizeof...(Ts)>());
+    }
+
+    auto await_resume() {
+        assert(_result != nullptr);
+        return std::move(_result->value());
```

Comment:
Why we don't return value with callbacks for vectors but returning here?

Reply:
variadic version no need index, because the sequence is the index. the index for vector version.